tools/distpartitioning/multi_dev_init.py

```diff
@@ -0,0 +1,499 @@
+    #pair is removed after the data communication
+    node_data[constants.NTYPE_ID] = rcvd_node_data[:,0]
+    node_data[constants.GLOBAL_TYPE_NID] = rcvd_node_data[:,1]
+    node_data[constants.GLOBAL_NID] = rcvd_node_data[:,2]
+    node_data.pop(constants.OWNER_PROCESS)
+
+def exchange_edge_data(rank, world_size, edge_data):
+    """
+    Exchange edge_data among processes in the world.
+    Prepare list of sliced data targeting each process and trigger
+    alltoallv_cpu to trigger messaging api
+
+    Parameters:
+    -----------
+    rank : int
+        rank of the process
+    world_size : int
+        total no. of processes
+    edge_data : dictionary
+        edge information, as a dicitonary which stores column names as keys and values
+        as column data. This information is read from the edges.txt file.
+    """
+    input_list = []
+    send_sizes = []
+    recv_sizes = []
+    start = timer()
+    for i in np.arange(world_size):
+        send_idx = (edge_data[constants.OWNER_PROCESS] == i)
+        send_idx = send_idx.reshape(edge_data[constants.GLOBAL_SRC_ID].shape[0])
+        filt_data = np.column_stack((edge_data[constants.GLOBAL_SRC_ID][send_idx == 1], \
+                                    edge_data[constants.GLOBAL_DST_ID][send_idx == 1], \
+                                    edge_data[constants.GLOBAL_TYPE_EID][send_idx == 1], \
+                                    edge_data[constants.ETYPE_ID][send_idx == 1], \
+                                    edge_data[constants.GLOBAL_EID][send_idx == 1]))
+        if(filt_data.shape[0] <= 0):
+            input_list.append(torch.empty((0,), dtype=torch.int64))
+            send_sizes.append(torch.empty((0,), dtype=torch.int64))
+        else:
+            input_list.append(torch.from_numpy(filt_data))
+            send_sizes.append(torch.tensor(filt_data.shape, dtype=torch.int64))
+        recv_sizes.append(torch.zeros((2,), dtype=torch.int64))
+    end = timer()
+    
+    dist.barrier ()
+    start = timer()
+    alltoall_cpu(rank, world_size, recv_sizes, send_sizes)
+    output_list = []
+    for s in recv_sizes: 
+        output_list.append(torch.zeros(s.tolist(), dtype=torch.int64))
+
+    dist.barrier ()
+    alltoallv_cpu(rank, world_size, output_list, input_list)
+    end = timer()
+    print('[Rank: ', rank, '] Time to send/rcv edge data: ', timedelta(seconds=end-start))
+
+    #Replace the values of the edge_data, with the received data from all the other processes.
+    rcvd_edge_data = torch.cat(output_list).numpy()
+    edge_data[constants.GLOBAL_SRC_ID] = rcvd_edge_data[:,0]
+    edge_data[constants.GLOBAL_DST_ID] = rcvd_edge_data[:,1]
+    edge_data[constants.GLOBAL_TYPE_EID] = rcvd_edge_data[:,2]
+    edge_data[constants.ETYPE_ID] = rcvd_edge_data[:,3]
+    edge_data[constants.GLOBAL_EID] = rcvd_edge_data[:,4]
+    edge_data.pop(constants.OWNER_PROCESS)
+
+
+def exchange_node_features(rank, world_size, node_data, node_features, ntypes_map, \
```

Comment:
why can i not find `exchange_edge_features`?

Reply:
This code is not instrumented yet. Per one of our discussions, we first implement only exchanging node features because there was no dataset which has edge features and this was never used before. So there is no way to test it on a real dataset.